[PYTORCH] Support max_pool2d_with_indices

trevor-m opened the pull request:
Pytorch converter doesn't gracefully handle the situation when a pytorch op has multiple outputs but the converter function only returns a single output.

* Fix specific case I encountered for `aten::max_pool2d_with_indices` by adding converter which returns a second dummy output.
* Also, improve handling of the situation by using `outputs[_get_output_name(op_node)] = relay_out` instead of `outputs[node_name] = relay_out` so that the assert in `_get_output_name` will be hit instead of a confusing key error later on. When a pytorch op has multiple outputs, node_name is a concatenation of all of the output names so downstream operators would never be able to find the specific output in `outputs` that they wanted.

masahi commented:
I don't think this is a bug. For your case, the op converter can return dummy result as the second output. See adaptive max pool converter below.

https://github.com/apache/incubator-tvm/blob/master/python/tvm/relay/frontend/pytorch.py#L602

trevor-m commented:
@masahi I see, thank you! I will change this PR to instead modify the converter for `aten::max_pool2d_with_indices` to have a second dummy output.

trevor-m commented:
@masahi I added a converter for max_pool_2d_with_indices.

I also changed the key in `outputs` to be `_get_output_name(op_node)` instead of `node_name`. This way, if a converter doesn't return the right number of outputs, the assert in `_get_output_name` will catch it instead right away of getting a key error when an op further down tries to access that output.

masahi commented:
@trevor-m Please update the PR title, and add a test for the new op converter in pytorch/test_forward.py.

You can add a test for `with_indices` variant in the existing max pool2d test, see
 https://github.com/apache/incubator-tvm/blob/master/tests/python/frontend/pytorch/test_forward.py#L523

zhiics commented:
@masahi please take another look. Thanks.

masahi commented on python/tvm/relay/frontend/pytorch.py:
```diff
@@ -2241,7 +2247,7 @@ def convert_operators(operators, outputs, ret_names, convert_map, prelude):
                 outputs.update(zip(out_names, relay_out))
             else:
-                outputs[node_name] = relay_out
+                outputs[_get_output_name(op_node)] = relay_out
```
I think it is clearer to directly assert `op_node.outputsSize() == 1` here, instead of calling `_get_output_name` again.

trevor-m replied:
Thanks, done!

masahi commented:
Sorry I made another last minute comment.

masahi commented:
@trevor-m Can you have a look at the failed test?

trevor-m commented:
Hi @masahi , the error is due to the dummy output. For some reason there is a `prim::TupleConstruct` operator in the pytorch graph which creates a relay.Tuple with the maxpool2d op and the dummy `None` output. The dummy `None` output causes the failure in relay.

masahi commented:
So `MaxPool` with `return_indieces=True` returns a tuple? Strange, it doesn't happen with adaptive pool. If you add another Torch module after max pool, which consumes the first output of max pool and ignores indieces, does this remove tuple creation?

trevor-m commented:
> So `MaxPool` with `return_indieces=True` returns a tuple? Strange, it doesn't happen with adaptive pool. If you add another Torch module after max pool, which consumes the first output of max pool and ignores indieces, does this remove tuple creation?

Yeah it appears to add the tuple. It looks like adaptive pool only returns indices when return_indices is True https://pytorch.org/docs/stable/nn.html#adaptivemaxpool3d which is why we don't see the issue with the adaptive pool tests. So the adaptive pool converter should actually only add the dummy output in those cases. But the behavior with zip() seems to not care about the extra None:
```
>>> list(zip(['1'], [5, None]))
[('1', 5)]
```
```

Anyway, I've updated the test to only consume the first output which solves the problem.

masahi commented:
Adaptive max pool in Torch jit always returns indices, regard less of `return_indices` param. This is the IR when running the adaptive max pool3d test (with `return_indices=False`):
```
graph(%self : __torch__.torch.nn.modules.pooling.AdaptiveMaxPool3d,
      %input : Float(1, 32, 16, 16, 16)):
  %2 : int = prim::Constant[value=1]() # /home/masa/projects/deep/pytorch/torch/nn/functional.py:837:0
  %3 : int = prim::Constant[value=1]() # /home/masa/projects/deep/pytorch/torch/nn/functional.py:837:0
  %4 : int = prim::Constant[value=1]() # /home/masa/projects/deep/pytorch/torch/nn/functional.py:837:0
  %5 : int[] = prim::ListConstruct(%2, %3, %4)
  %6 : Float(1, 32, 1, 1, 1), %7 : Long(1, 32, 1, 1, 1) = aten::adaptive_max_pool3d(%input, %5) # /home/masa/projects/deep/pytorch/torch/nn/functional.py:837:0
  return (%6)
```

But yeah, if I add `return_indices=True` to adaptive pool test, I get the same error.

masahi commented:
But still the "dummy output" solution is better then the one you proposed initially because it may happen that the second or other outputs do get used by other ops, so we should keep them.

masahi commented:
@trevor-m Can you try retrigger the CI? Seems it failed for unrelated reason.

masahi commented:
Thanks @trevor-m @zhiics
